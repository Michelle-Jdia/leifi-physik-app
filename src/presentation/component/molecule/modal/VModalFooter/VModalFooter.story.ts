import type { Meta, StoryFn } from '@storybook/vue3';
import VModalFooter from '@/molecule/modal/VModalFooter/VModalFooter.vue';
import VButton from '@/atom/button/VButton/VButton.vue';

export default {
    component: VModalFooter,
} as Meta<typeof VModalFooter>;

const Template: StoryFn<typeof VModalFooter> = (args) => ({
    components: {
        VModalFooter,
        VButton,
    },
    setup() {
        return {
            args,
        };
    },
    template: `
        <v-modal-footer v-bind="args" style="position: relative">
          <template #before>
            <button class="v-text-pink">Sammlung löschen</button>
          </template>
          <v-button color="tertiary">+ Artikel</v-button>

          <v-button modifier="v-ml-box-md" color="tertiary">+ Überschrift</v-button>
        </v-modal-footer>
    `,
});

export const Base = Template.bind({});

Base.args = {};
