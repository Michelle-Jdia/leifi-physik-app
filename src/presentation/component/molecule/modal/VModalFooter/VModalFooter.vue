<script setup lang="ts">
import type { ClassValue } from 'clsx';
import { cn } from '@/presentation/helper/style';

interface Props {
    modifier?: ClassValue;
}

defineProps<Props>();
</script>

<template>
    <div :class="cn('modal-footer', modifier)">
        <div v-if="$slots.before" class="v-bg-white v-p-box-xl">
            <slot name="before" />
        </div>

        <div class="v-flex v-border-t v-border-gray-50 v-bg-white v-p-box-xl">
            <slot />
        </div>
    </div>
</template>
