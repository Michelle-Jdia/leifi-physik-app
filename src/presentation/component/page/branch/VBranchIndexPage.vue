<script setup lang="ts">
import type { SliderItem } from '@/molecule/slider/VTopicBoxSlider/VTopicBoxSlider.vue';
import type { RefresherCustomEvent } from '@ionic/vue';
import type { ComputedRef } from 'vue';
import { onIonViewWillEnter } from '@ionic/vue';
import { computed } from 'vue';
import { createRefresherHandler } from '@/connection/helper/refresher';
import { RouteName } from '@/connection/router/routeName';
import { useBranchState } from '@/state/branchState';
import { navigateToSearch } from '@/store/searchStore';
import { createTopicSlider } from '@/presentation/helper/topic';
import VBranchHeader from '@/layout/header/VBranchHeader/VBranchHeader.vue';
import VPage from '@/layout/page/VPage.vue';
import VGlobalSearchInput from '@/container/form/VGlobalSearchInput/VGlobalSearchInput.vue';
import VTopicBoxSlider from '@/molecule/slider/VTopicBoxSlider/VTopicBoxSlider.vue';
import VBranchBox from '@/atom/box/VBranchBox/VBranchBox.vue';
import VLoader from '@/atom/loader/VLoader/VLoader.vue';
import VRefresher from '@/atom/refresher/VRefresher/VRefresher.vue';

type TopicSliderItem = Record<string, SliderItem[]>;

const branchState = useBranchState();

const sliders: ComputedRef<TopicSliderItem | void> = computed(() => {
    if (!branchState.topics) {
        return;
    }

    return branchState.topics.reduce((sliders: TopicSliderItem, topic) => {
        if (topic && topic[0]) {
            const branchId = topic[0].referenced_branch;

            if (!sliders[branchId]) {
                sliders[branchId] = topic.map(createTopicSlider);
            }
        }

        return sliders;
    }, {});
});

async function handleRefresh(event: RefresherCustomEvent): Promise<void> {
    await createRefresherHandler({
        event: event,
        callback: async () => {
            await branchState.getLiveTopicsAndBranches();
            await branchState.getTopics();
        },
    });
}

onIonViewWillEnter(branchState.getTopics);
</script>

<template>
    <v-page>
        <v-loader
            :logs="['useBranches', 'getTopicsByBranch', 'useBranchesLive', 'getTopicsByBranchLive']"
        />

        <template #header>
            <v-branch-header @on-submit="navigateToSearch">Alle Inhalte</v-branch-header>
        </template>

        <template #refresher>
            <v-refresher @ion-refresh="handleRefresh" />
        </template>

        <v-global-search-input modifier="v-mb-box-xl md:v-hidden" />

        <h1 class="v-h2 v-font-bold">Teilgebiete der Physik</h1>

        <div class="v-mt-box-xl v-grid v-grid-cols-2 v-gap-box-md md:v-grid-cols-4">
            <template v-for="branch in branchState.branches" :key="branch.name">
                <router-link :to="{ name: RouteName.BRANCH_ID, params: { id: branch.id } }">
                    <v-branch-box
                        :img="{
                            src: branch.icon?.url || '',
                            alt: branch.icon?.alt || '',
                        }"
                    >
                        {{ branch.name }}
                    </v-branch-box>
                </router-link>
            </template>
        </div>

        <v-topic-box-slider
            v-for="branch in branchState.branches"
            :key="branch.id"
            :color="branch.color"
            modifier="v-mt-box-md"
            :branch-img="{
                src: branch.icon?.url || '',
                alt: branch.icon?.alt || '',
            }"
            :link="{ name: RouteName.BRANCH_ID, params: { id: branch.id } }"
            :slider-items="(sliders && sliders[branch.id]) || []"
        >
            {{ branch.name }}
        </v-topic-box-slider>
    </v-page>
</template>
