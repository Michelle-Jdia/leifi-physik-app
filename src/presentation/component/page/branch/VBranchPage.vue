<script setup lang="ts">
import type { SliderItem } from '@/molecule/slider/VTopicBoxSlider/VTopicBoxSlider.vue';
import type { RefresherCustomEvent } from '@ionic/vue';
import type { ComputedRef } from 'vue';
import { computed } from 'vue';
import { createRefresherHandler } from '@/connection/helper/refresher';
import { useBranch } from '@/store/branchStore';
import { navigateToSearch } from '@/store/searchStore';
import {
    getTopicsByBranchLive,
    isTopicFirstDegree,
    isTopicSecondDegree,
    isTopicThirdDegree,
    TopicDegree,
    useTopicsByBranch,
} from '@/store/topicStore';
import { createTopicSlider } from '@/presentation/helper/topic';
import VBranchHeader from '@/layout/header/VBranchHeader/VBranchHeader.vue';
import VPage from '@/layout/page/VPage.vue';
import VTopicBoxSlider from '@/molecule/slider/VTopicBoxSlider/VTopicBoxSlider.vue';
import VLoader from '@/atom/loader/VLoader/VLoader.vue';
import VRefresher from '@/atom/refresher/VRefresher/VRefresher.vue';

interface Props {
    id: string;
}

const props = defineProps<Props>();

type TopicSliderItem = Record<string, SliderItem[]>;

const themeTitle = 'Themenbereiche';

const { data: branch, getData: getBranch } = useBranch({
    params: {
        id: props.id,
    },
});

const { data: topics, getData: getTopics } = useTopicsByBranch({
    params: {
        branchId: props.id,
    },
});

const { getData: getLiveTopics } = getTopicsByBranchLive({
    params: {
        branchId: props.id,
    },
});

const sliders: ComputedRef<TopicSliderItem | void> = computed(() => {
    if (!topics.value) {
        return;
    }

    const levelOneTopics = topics.value.filter(isTopicFirstDegree);
    const levelTwoTopics = topics.value.filter(isTopicSecondDegree);
    const levelThreeTopics = topics.value.filter(isTopicThirdDegree);

    return {
        levelOne: levelOneTopics.map(createTopicSlider),
        levelTwo: levelTwoTopics.map(createTopicSlider),
        levelThree: levelThreeTopics.map(createTopicSlider),
    };
});

function getTopicHeadline(key: string): TopicDegree | void {
    if (key === 'levelOne') {
        return TopicDegree.one;
    }

    if (key === 'levelTwo') {
        return TopicDegree.two;
    }

    if (key === 'levelThree') {
        return TopicDegree.three;
    }
}

async function handleRefresh(event: RefresherCustomEvent): Promise<void> {
    await createRefresherHandler({
        event: event,
        callback: async () => {
            await getLiveTopics();
            await getBranch();
            await getTopics();
        },
    });
}

getBranch();
getTopics();
</script>

<template>
    <v-page>
        <v-loader :logs="['useBranch', 'getTopicsByBranch']" />

        <template #header>
            <v-branch-header @on-submit="navigateToSearch">
                <template v-if="branch">
                    {{ branch.name }}
                </template>

                <template #sub-title>Teilgebiete</template>
            </v-branch-header>
        </template>

        <template #refresher>
            <v-refresher @ion-refresh="handleRefresh" />
        </template>

        <template v-if="sliders && branch">
            <template v-for="key in Object.keys(sliders)" :key="key">
                <div v-if="sliders[key] && sliders[key]?.length" class="v-mb-section">
                    <h2 class="v-h2">{{ getTopicHeadline(key) }}</h2>

                    <v-topic-box-slider
                        v-if="sliders[key]"
                        :key="branch.id"
                        modifier="v-mt-box-md"
                        :color="branch.color"
                        :branch-img="{
                            src: branch.icon?.url || '',
                            alt: branch.icon?.alt || '',
                        }"
                        :slider-items="sliders[key] || []"
                    >
                        {{ themeTitle }}
                    </v-topic-box-slider>
                </div>
            </template>
        </template>
    </v-page>
</template>
