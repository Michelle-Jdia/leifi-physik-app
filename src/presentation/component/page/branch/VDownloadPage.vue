<script setup lang="ts">
import type { Collection } from '@/data/type/app/collection';
import type { RefresherCustomEvent } from '@ionic/vue';
import type { Ref } from 'vue';
import { onIonViewWillEnter } from '@ionic/vue';
import { computed, ref, watch } from 'vue';
import { $t } from '@/application/translation';
import { ContentType } from '@/data/type/helper/contentType';
import { useDownloadTopicBranch } from '@/connection/helper/download';
import { createRefresherHandler } from '@/connection/helper/refresher';
import { RouteName } from '@/connection/router/routeName';
import { useArticleState } from '@/state/articleState';
import { findArticleInCollection } from '@/store/collectionStore';
import { useDownloadLive } from '@/store/downloadStore';
import { addDownloadToHistory } from '@/store/historyStore';
import { handleLinksClick } from '@/presentation/helper/handleLink';
import { resizeBrowser } from '@/presentation/helper/swiffy';
import VArticleHeader from '@/layout/header/VArticleHeader.vue';
import VPage from '@/layout/page/VPage.vue';
import VInputButton from '@/molecule/form/VInputButton/VInputButton.vue';
import VAccordion from '@/atom/accordion/VAccordion.vue';
import VButton from '@/atom/button/VButton/VButton.vue';
import VCheckbox from '@/atom/checkbox/VCheckbox/VCheckbox.vue';
import VDownloadContentWrapper from '@/atom/contentWrapper/VDownloadContentWrapper.vue';
import VIcon from '@/atom/icon/VIcon/VIcon.vue';
import VLoader from '@/atom/loader/VLoader/VLoader.vue';
import VModal from '@/atom/modal/VModal/VModal.vue';
import VRefresher from '@/atom/refresher/VRefresher/VRefresher.vue';
import VAnimationScript from '@/atom/script/VAnimationScript/VAnimationScript.vue';
import VBundle from '@/atom/script/VBundle/VBundle.vue';
import VH5p from '@/atom/script/VH5p/VH5p.vue';

interface Props {
    id: string;
}
interface DownloadFile {
    id: string;
    url: string;
    size: string;
}

const props = defineProps<Props>();
const isSwiffyReady: Ref<boolean> = ref(false);
const articleState = useArticleState();
const addToCollectionsModal = ref();
const createCollectionModal = ref();
const rawHtmlLinksContainer = ref();
const tempCollectionsToUpdate: Collection[] = [];
const { data: pageData, getData: getDownloadTopicBranch } = useDownloadTopicBranch(props.id);

const { getData: getDownloadLive } = useDownloadLive({
    params: {
        id: props.id,
    },
});

function openCollectionModal(): void {
    articleState.getCollections();
    tempCollectionsToUpdate.length = 0;
    addToCollectionsModal.value?.$el.present();
}

function toggleArticleInTempList(event: CustomEvent, item: Collection): void {
    if (event.detail.checked === true) {
        tempCollectionsToUpdate.push(item);
    }

    if (event.detail.checked === false) {
        const filtered = tempCollectionsToUpdate.filter(
            (collection) => !(collection.id === item.id && collection.type === item.type),
        );

        tempCollectionsToUpdate.length = 0;
        tempCollectionsToUpdate.push(...filtered);
    }
}

function handleModalConfirmation(): void {
    if (!pageData.value || !pageData.value.download) {
        return;
    }

    articleState.saveArticlesToCollections(pageData.value.download, tempCollectionsToUpdate);
}

function handleCreateCollectionSubmit({ value }: { value: string }): void {
    if (!pageData.value || !pageData.value.download || !value) {
        return;
    }

    articleState.createAndAddArticleToNewCollection(pageData.value.download, value);
}

function openCreateCollectionModal(): void {
    // addToCollectionsModal.value?.$el.dismiss();
    createCollectionModal.value?.$el.present();
}

function setSwiffiyToReady() {
    isSwiffyReady.value = true;
}

function getFileSizeInBites(size: number): string {
    const sizes = ['B', 'KB', 'MB', 'GB'];
    const kb = 1024;
    let i = 0;

    do {
        size = size / kb;
        i++;
    } while (size >= kb);

    return `${size.toFixed(2)} ${sizes[i]}`;
}

const downloadedFiles = computed<DownloadFile[]>(() => {
    if (!pageData.value || !pageData.value.download) {
        return [];
    }

    return pageData.value.download.download_file.map((file) => ({
        id: file.id,
        size: `Größe: ${getFileSizeInBites(file.size)}`,
        url: file.url,
    }));
});

async function handleRefresh(event: RefresherCustomEvent): Promise<void> {
    await createRefresherHandler({
        event: event,
        callback: async () => {
            await getDownloadLive();
            await getDownloadTopicBranch();
        },
    });
}

watch(rawHtmlLinksContainer, () => {
    if (!rawHtmlLinksContainer.value) {
        return;
    }

    handleLinksClick(rawHtmlLinksContainer.value);
});

onIonViewWillEnter(async () => {
    addDownloadToHistory(props.id);

    await getDownloadTopicBranch();

    resizeBrowser();
});
</script>

<template>
    <v-page modifier="v-pt-0 md:v-pt-0">
        <template v-if="pageData">
            <v-bundle />

            <v-h5p />

            <v-animation-script :onload="setSwiffiyToReady" />
        </template>

        <v-loader :logs="['useDownload']" />

        <template #refresher>
            <v-refresher @ion-refresh="handleRefresh" />
        </template>

        <template #header>
            <v-article-header
                :id="id"
                :type="ContentType.DOWNLOAD"
                @on-open-add-to-collection="openCollectionModal"
            >
                <template v-if="pageData && pageData.topic && pageData.branch" #pre-title>
                    <router-link
                        :to="{ name: RouteName.BRANCH_ID, params: { id: pageData.branch.id } }"
                    >
                        {{ pageData.branch.name }}
                    </router-link>
                    /
                    <router-link
                        :to="{ name: RouteName.TOPIC_ID, params: { id: pageData.topic.id } }"
                    >
                        {{ pageData.topic.name }}
                    </router-link>
                </template>

                <template v-if="pageData && pageData.download">
                    {{ pageData.download.title }}
                </template>
            </v-article-header>
        </template>

        <template v-if="pageData">
            <h1 class="v-pt-box-xl v-text-h2">{{ $t.download.headline }}</h1>

            <p class="v-p v-mt-box-lg">
                <span class="v-font-bold">Typ:</span>

                {{ pageData.download.download_type }}
            </p>

            <div
                v-if="isSwiffyReady"
                class="site v-mt-box-xl v-border-t v-border-t-gray-80 v-pt-box-xl"
            >
                <div v-html="pageData.download.download_description" />

                <div
                    v-if="pageData.download.animation_code?.rendered_string"
                    v-html="pageData.download.animation_code.rendered_string"
                />

                <template v-if="pageData.download.animation_code?.attachments">
                    <component
                        :is="attachment.tag"
                        v-for="(attachment, indexAttachment) in pageData.download.animation_code
                            .attachments"
                        :key="`${indexAttachment}-task-question-attachment`"
                        :type="attachment.type"
                    >
                        {{ attachment.content }}
                    </component>
                </template>
            </div>

            <div class="site">
                <template v-if="downloadedFiles.length">
                    <template v-for="file in downloadedFiles" :key="file.id">
                        <p class="v-p v-mt-box-lg">{{ file.size }}</p>

                        <a v-if="file.url" :href="file.url" target="_blank">
                            <v-button class="v-w-full" size="small" :is-rounded="false">{{
                                $t.download.fielDownloadButtonLabel
                            }}</v-button>
                        </a>
                    </template>
                </template>
            </div>

            <v-accordion v-if="pageData.download.download_didactic_info.trim().length">
                <template #label>{{ $t.download.accordionLabel }}:</template>

                <v-download-content-wrapper>
                    <div v-html="pageData.download.download_didactic_info" />
                </v-download-content-wrapper>
            </v-accordion>
        </template>

        <v-modal ref="addToCollectionsModal" with-save @on-confirm="handleModalConfirmation">
            <template #title>{{ $t.collection.addCollection }}</template>

            <v-button modifier="v-w-full" @click="openCreateCollectionModal">
                <v-icon src="icon/add-collection.svg" with-space-right />

                {{ $t.collection.createCollection }}
            </v-button>

            <template v-if="articleState.collections && articleState.collections.length">
                <div
                    v-for="collection in articleState.collections"
                    :key="collection.id"
                    class="v-mt-box-md v-flex v-items-center v-rounded v-border v-border-gray-20 v-bg-white v-p-box"
                >
                    <v-checkbox
                        :disabled="
                            findArticleInCollection(collection, props.id, ContentType.DOWNLOAD)
                        "
                        :checked="
                            findArticleInCollection(collection, props.id, ContentType.DOWNLOAD)
                        "
                        modifier="v-mr-auto v-w-full v-h3 v-font-bold"
                        @ion-change="(event) => toggleArticleInTempList(event, collection)"
                    >
                        {{ collection.private_title }}
                    </v-checkbox>
                </div>
            </template>
        </v-modal>

        <v-modal ref="createCollectionModal">
            <template #title>{{ $t.collection.addCollection }}</template>

            <v-input-button
                icon="icon/add-collection.svg"
                :button-text="$t.collection.create"
                :placeholder="$t.collection.collectionName"
                @on-submit="handleCreateCollectionSubmit"
            />
        </v-modal>
    </v-page>
</template>
