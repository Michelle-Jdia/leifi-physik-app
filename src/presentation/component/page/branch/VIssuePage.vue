<script setup lang="ts">
import type { Collection } from '@/data/type/app/collection';
import type { Paragraph } from '@/data/type/app/field/Paragraph';
import type { Task } from '@/data/type/app/task';
import type { IonModal, RefresherCustomEvent } from '@ionic/vue';
import type { ComputedRef, Ref } from 'vue';
import { onIonViewWillEnter } from '@ionic/vue';
import { computed, ref, watch } from 'vue';
import { $t } from '@/application/translation';
import {
    ContentType,
    isParagraphBoxExtended,
    isParagraphHeadline,
    isParagraphToc,
} from '@/data/type/helper/contentType';
import { useIssueTopicBranch } from '@/connection/helper/issue';
import { closeModal, openModal } from '@/connection/helper/modal';
import { getContextualRouteNameByType } from '@/connection/helper/navigation';
import { createRefresherHandler } from '@/connection/helper/refresher';
import { RouteName } from '@/connection/router/routeName';
import { useArticleState } from '@/state/articleState';
import { useIssueState } from '@/state/issueState';
import { findArticleInCollection } from '@/store/collectionStore';
import { addIssueToHistory } from '@/store/historyStore';
import { categorySingularMap, useIssueLive } from '@/store/issueStore';
import { sortTasksByType, useTasksByIssue, useTasksByIssueLive } from '@/store/taskStore';
import { handleLinksClick } from '@/presentation/helper/handleLink';
import { resizeBrowser } from '@/presentation/helper/swiffy';
import VArticleHeader from '@/layout/header/VArticleHeader.vue';
import VPage from '@/layout/page/VPage.vue';
import VInputButton from '@/molecule/form/VInputButton/VInputButton.vue';
import VButton from '@/atom/button/VButton/VButton.vue';
import VCheckbox from '@/atom/checkbox/VCheckbox/VCheckbox.vue';
import VIssueParagraphWrapper from '@/atom/contentWrapper/VIssueParagraphWrapper.vue';
import VIcon from '@/atom/icon/VIcon/VIcon.vue';
import VLinkItem from '@/atom/item/VLinkItem/VLinkItem.vue';
import VLoader from '@/atom/loader/VLoader/VLoader.vue';
import VModal from '@/atom/modal/VModal/VModal.vue';
import VBoxExtendedParagraph from '@/atom/paragraph/VBoxExtendedParagraph.vue';
import VRefresher from '@/atom/refresher/VRefresher/VRefresher.vue';
import VAnimationScript from '@/atom/script/VAnimationScript/VAnimationScript.vue';
import VBundle from '@/atom/script/VBundle/VBundle.vue';
import VH5p from '@/atom/script/VH5p/VH5p.vue';

interface Props {
    id: string;
}
type IonModalInstance = InstanceType<typeof IonModal>;

const props = defineProps<Props>();
const articleState = useArticleState();
const addToCollectionsModal: Ref<IonModalInstance | undefined> = ref();
const createCollectionModal: Ref<IonModalInstance | undefined> = ref();
const tempCollectionsToUpdate: Collection[] = [];
const isSwiffyReady = ref(false);
const rawHtmlLinksContainer = ref();
const { data: pageData, getData: getIssueTopicBranch } = useIssueTopicBranch(props.id);

const { data: tasks, getData: getTasksByIssue } = useTasksByIssue({
    params: {
        issueId: props.id,
    },
});

const { getData: getTasksByIssueLive } = useTasksByIssueLive({
    params: {
        issueId: props.id,
    },
});

const { getData: getIssueLive } = useIssueLive({
    params: {
        id: props.id,
    },
});

const issueState = useIssueState();

const tasksByType: ComputedRef<Record<string, Task[]> | void> = computed(() => {
    if (!tasks.value || !tasks.value.length) {
        return;
    }

    return sortTasksByType(tasks.value);
});

const categoryHeadline: ComputedRef<string> = computed(() => {
    const name = pageData.value?.issue.issue_category.name;

    if (!name) {
        return '';
    }

    return categorySingularMap.get(name) || name;
});

const refAnchors: Ref<HTMLHeadingElement[]> = ref([]);

const addRefAnchor = (index: number) => (anchor: any) => {
    // we need to use index here, otherwise the loops could spam multiple refs in the array
    refAnchors.value[index] = anchor;
};

function openCollectionModal(): void {
    articleState.getCollections();
    tempCollectionsToUpdate.length = 0;
    openModal(addToCollectionsModal.value);
}

function toggleArticleInTempList(event: CustomEvent, item: Collection): void {
    if (event.detail.checked === true) {
        tempCollectionsToUpdate.push(item);
    }

    if (event.detail.checked === false) {
        const filtered = tempCollectionsToUpdate.filter(
            (collection) => !(collection.id === item.id && collection.type === item.type),
        );

        tempCollectionsToUpdate.length = 0;
        tempCollectionsToUpdate.push(...filtered);
    }
}

function handleModalConfirmation(): void {
    if (!pageData.value || !pageData.value.issue) {
        return;
    }

    articleState.saveArticlesToCollections(pageData.value.issue, tempCollectionsToUpdate);
}

function handleCreateCollectionSubmit({ value }: { value: string }): void {
    if (!pageData.value || !pageData.value.issue || !value) {
        return;
    }

    articleState.createAndAddArticleToNewCollection(pageData.value.issue, value);
    closeModal(createCollectionModal.value);
    closeModal(addToCollectionsModal.value);
}

function scrollToElement(element: HTMLHeadingElement) {
    element.scrollIntoView({
        behavior: 'smooth',
    });
}

const refHeadlines: Ref<HTMLHeadingElement[]> = ref([]);

const filteredRefHeadlines: ComputedRef<HTMLHeadingElement[] | undefined> = computed(() => {
    return refHeadlines.value.filter((headline) => headline);
});

const addRefHeadlines = (index: number, paragraph: Paragraph) => (element: any) => {
    if (
        !element ||
        !element.children.length ||
        !element.children[0] ||
        !isParagraphHeadline(paragraph)
    ) {
        return;
    }

    /* add the branch color class directly to the element */
    element.children[0].classList.add(headlineParagraphStyle.value);
    refHeadlines.value[index] = element.children[0];
};

const headlineParagraphStyle: ComputedRef<string> = computed(() => {
    return pageData.value?.branch.color || '';
});

function setSwiffiyToReady() {
    isSwiffyReady.value = true;
}

addIssueToHistory(props.id);
getTasksByIssue();

async function handleRefresh(event: RefresherCustomEvent): Promise<void> {
    await createRefresherHandler({
        event: event,
        callback: async () => {
            await getIssueLive();
            await getTasksByIssueLive();
            await getIssueTopicBranch();
            await getTasksByIssue();
        },
    });
}

watch(rawHtmlLinksContainer, () => {
    if (!rawHtmlLinksContainer.value) {
        return;
    }

    handleLinksClick(rawHtmlLinksContainer.value);
});

onIonViewWillEnter(async () => {
    issueState.getPrevNextIssueIds(props.id);

    await getIssueTopicBranch();

    resizeBrowser();
});
</script>

<template>
    <v-page modifier="v-pt-0 md:v-pt-0">
        <template v-if="pageData">
            <v-bundle />

            <v-h5p />

            <v-animation-script :onload="setSwiffiyToReady" />
        </template>

        <v-loader :logs="['useIssue']" />

        <template #refresher>
            <v-refresher @ion-refresh="handleRefresh" />
        </template>

        <template #header>
            <v-article-header
                :id="id"
                :anchors="refAnchors"
                :type="ContentType.ISSUE"
                @on-open-add-to-collection="openCollectionModal"
            >
                <template v-if="pageData && pageData.topic && pageData.branch" #pre-title>
                    <router-link
                        :to="{ name: RouteName.BRANCH_ID, params: { id: pageData.branch.id } }"
                    >
                        {{ pageData.branch.name }}
                    </router-link>
                    /
                    <router-link
                        :to="{ name: RouteName.TOPIC_ID, params: { id: pageData.topic.id } }"
                    >
                        {{ pageData.topic.name }}
                    </router-link>
                </template>

                <template v-if="pageData && pageData.issue">
                    {{ pageData.issue.title }}
                </template>
            </v-article-header>
        </template>

        <template v-if="pageData?.issue">
            <h1 :ref="addRefAnchor(0)" class="v-pt-box-xl v-text-h2">
                {{ categoryHeadline }}
            </h1>

            <h2 v-if="pageData.issue.issue_teaser_title" class="v-mt-box-lg v-text-h4">
                {{ pageData.issue.issue_teaser_title }}
            </h2>

            <div
                v-if="pageData && pageData.issue.issue_teaser_content.trim().length"
                class="site v-mt-box-xl"
            >
                <p
                    ref="rawHtmlLinksContainer"
                    class="v-p"
                    v-html="pageData.issue.issue_teaser_content"
                />
            </div>

            <div
                v-if="isSwiffyReady"
                class="site v-mt-box-xl v-border-t v-border-t-gray-80 v-pt-box-xl"
            >
                <div
                    v-for="(paragraph, index) in pageData?.issue.issue_paragraphs"
                    :key="`${index}-issue-paragraph`"
                >
                    <div v-if="!isParagraphBoxExtended(paragraph)">
                        <v-issue-paragraph-wrapper v-if="!isParagraphToc(paragraph)">
                            <div ref="rawHtmlLinksContainer" v-html="paragraph.rendered_string" />

                            <component
                                :is="attachment.tag"
                                v-for="(attachment, indexAttachment) in paragraph.attachments"
                                :key="`${indexAttachment}-issue-attachment`"
                                :type="attachment.type"
                            >
                                {{ attachment.content }}
                            </component>
                        </v-issue-paragraph-wrapper>

                        <div v-if="isParagraphToc(paragraph)">
                            <p
                                v-for="(headline, i) in filteredRefHeadlines"
                                :key="`${i}-issue-paragraph-headline`"
                                @click="scrollToElement(headline)"
                            >
                                <i class="fa-solid fa-chevron-right v-mr-2" />
                                {{ headline.innerText }}
                            </p>
                        </div>
                    </div>

                    <v-issue-paragraph-wrapper v-if="isParagraphBoxExtended(paragraph)">
                        <v-box-extended-paragraph :content="paragraph" />
                    </v-issue-paragraph-wrapper>
                </div>
            </div>

            <div
                v-if="isSwiffyReady && pageData?.issue.issue_paragraphs_ext.length"
                class="site v-mt-box-md"
            >
                <div
                    v-for="(paragraph_ext, index) in pageData?.issue.issue_paragraphs_ext"
                    :key="`${index}-issue-paragraph_ext`"
                >
                    <v-issue-paragraph-wrapper v-if="!isParagraphBoxExtended(paragraph_ext)">
                        <div
                            :ref="addRefHeadlines(index, paragraph_ext)"
                            class="v-overflow-scroll"
                            v-html="paragraph_ext.rendered_string"
                        />
                    </v-issue-paragraph-wrapper>

                    <v-issue-paragraph-wrapper v-if="isParagraphBoxExtended(paragraph_ext)">
                        <v-box-extended-paragraph :content="paragraph_ext" />
                    </v-issue-paragraph-wrapper>
                </div>
            </div>
        </template>

        <v-loader is-relative :logs="['useTasksByIssue']" />

        <div v-if="tasksByType" class="v-mt-section-md">
            <h2 :ref="addRefAnchor(1)" class="v-h2 v-pt-box-xl">
                {{ $t.issue.taskSectionHeadline }}
            </h2>

            <div v-for="(sortedTasks, type) in tasksByType" :key="type">
                <h3 class="v-h4 v-mt-box-xl">{{ type }}</h3>

                <div v-for="task in sortedTasks" :key="task.id" class="v-mt-box-sm">
                    <router-link
                        :to="{
                            name: getContextualRouteNameByType(task.type),
                            params: { id: task.id },
                        }"
                    >
                        <v-link-item :level="task.level.color">{{ task.title }}</v-link-item>
                    </router-link>
                </div>
            </div>
        </div>

        <div v-if="pageData?.issue" class="v-mt-section v-flex">
            <v-loader class="v-w-full" is-relative :logs="['usePrevIssue', 'useNextIssue']" />

            <v-button
                v-if="issueState.prevIssueId"
                modifier="v-capitalize v-w-full"
                :router-link="{
                    name: getContextualRouteNameByType(ContentType.ISSUE),
                    params: { id: issueState.prevIssueId },
                }"
            >
                {{ $t.issue.buttonPrevIssue }}
            </v-button>

            <v-button
                v-if="issueState.nextIssueId"
                modifier="v-capitalize v-w-full"
                :router-link="{
                    name: getContextualRouteNameByType(ContentType.ISSUE),
                    params: { id: issueState.nextIssueId },
                }"
            >
                {{ $t.issue.buttonNextIssue }}
            </v-button>
        </div>

        <v-modal ref="addToCollectionsModal" with-save @on-confirm="handleModalConfirmation">
            <template #title>{{ $t.collection.addCollection }}</template>

            <v-button modifier="v-w-full" @click="openModal(createCollectionModal)">
                <v-icon src="icon/add-collection.svg" with-space-right />

                {{ $t.collection.createCollection }}
            </v-button>

            <template v-if="articleState.collections && articleState.collections.length">
                <div
                    v-for="collection in articleState.collections"
                    :key="collection.id"
                    class="v-mt-box-md v-flex v-items-center v-rounded v-border v-border-gray-20 v-bg-white v-p-box"
                >
                    <v-checkbox
                        :disabled="findArticleInCollection(collection, props.id, ContentType.ISSUE)"
                        :checked="findArticleInCollection(collection, props.id, ContentType.ISSUE)"
                        modifier="v-mr-auto v-w-full v-h3 v-font-bold"
                        @ion-change="(event) => toggleArticleInTempList(event, collection)"
                    >
                        {{ collection.private_title }}
                    </v-checkbox>
                </div>
            </template>
        </v-modal>

        <v-modal ref="createCollectionModal">
            <template #title>{{ $t.collection.addCollection }}</template>

            <v-input-button
                icon="icon/add-collection.svg"
                :button-text="$t.collection.create"
                :placeholder="$t.collection.collectionName"
                @on-submit="handleCreateCollectionSubmit"
            />
        </v-modal>
    </v-page>
</template>
