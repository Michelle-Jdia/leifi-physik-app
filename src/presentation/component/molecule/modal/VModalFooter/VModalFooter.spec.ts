import { describe } from 'vitest';
import { testSnapshot } from '@/presentation/helper/test';
import { branch } from '@/presentation/static/branch';
import VModalFooter from '@/molecule/modal/VModalFooter/VModalFooter.vue';

describe('VModalFooter', () => {
    const options = {
        slots: {
            default: branch.title,
        },
    };

    testSnapshot(VModalFooter, options);
});
