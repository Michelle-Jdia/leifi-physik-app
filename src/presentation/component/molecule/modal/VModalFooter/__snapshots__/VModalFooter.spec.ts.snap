// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`VModalFooter > snapshot 1`] = `
<div
  class="modal-footer"
>
  <!--v-if-->
  <div
    class="v-flex v-border-t v-border-gray-50 v-bg-white v-p-box-xl"
  >
    
    Mechanik
    
  </div>
</div>
`;
